### ISSUE TITLE
Allow the user to read the location of app configuration directory

### ISSUE THREAD

**Issue by xgfone** [fyne-io/fyne#1109]
### Is your feature request related to a problem? Please describe:

Yes. For the default implementation of `fyne.App`, it uses the fixed path to store the settings and preferences by the different OS, like this,
```go
func rootConfigDir() string {
	homeDir, _ := os.UserHomeDir()

	desktopConfig := filepath.Join(homeDir, ".config")
	return filepath.Join(desktopConfig, "fyne")
}
```

For settings, the path of its configuration file is that as following,
```go
// StoragePath returns the location of the settings storage
func (sc *SettingsSchema) StoragePath() string {
	return filepath.Join(rootConfigDir(), "settings.json")
}

func (s *settings) load() {
	err := s.loadFromFile(s.schema.StoragePath())
	if err != nil {
		fyne.LogError("Settings load error:", err)
	}

	s.setupTheme()
}
```

That's, for all kinds of fyne apps, those settings files will use the same one.  It may be not what we want.

I hope that there is an API to allow the user to customize it.

### Is it possible to construct a solution with the existing API?

Yes.

### Describe the solution you'd like to see:

Add a implicit global variable to save the setting of the user, and a exported function to modify it. If the variable is not empty, `rootConfigDir` returns it. Or return the one supplied by the default implementation. Like this
```go
var rootDir string

func SetRootConfigDir(dir string) {
	rootDir = dir
}

func rootConfigDir() string {
	if rootDir != "" {
		return rootDir
	}

	homeDir, _ := os.UserHomeDir()
	desktopConfig := filepath.Join(homeDir, ".config")
	return filepath.Join(desktopConfig, "fyne")
}
```

**Comment by andydotxyz** [comment]
I would like to understand this a little more. The storage locations have been chosen to place configuration in the correct place per Operating System specification.
Can you please expand on the reason to want to change this location? The rationale you provided above was "It may be not what we want." - can you be more specific?

**Comment by xgfone** [comment]
Our company want to use `Fyne` to develop a client, but some coworkers don't like `~/.config/fyne` or `%HOMEPATH%\AppData\AppData\fyne` as the setting directory and hope to use the directory which they assign.

Besides `setting` and `preferences`, moreover, we have other configurations to store. And we want to store them together, but the root configuration directory is implicit. Yes, it is supplied by `rootConfigDir`. Unless we check the specific directory for the specific platform from the sources, we don't known where they are saved. Oh, it's so dejected and helpless!

**Comment by andydotxyz** [comment]
Thanks. Is it possible to explain why your co-workers don't link the system specifid location?
You're quite right that the rootConfigDir is not exported - I was quite nervous about making an API to find those files in case they were edited manually.

There are app configurations but also global configuration values, if your app were to override where the root config was stored it would not (as things stand) be able to respect global values such as theme or user scale value because you have told it to ignore that global location.

I hope you can see this could be a can of worms - any more information about the rationale would be appreciated.

**Comment by xgfone** [comment]
> Is it possible to explain why your co-workers don't link the system specifid location?

Taking `Windows` as example, someone wants to place the configurations into `D:\somewhere`, not `C:\somewhere`; so he maybe have no need to reconfigure it after reinstalling OS. Moreover, we want to store other configurations in where `settings` and `preferences` are located. So we need to known where it's.

For linux, the root configuration directory is `~/.config/fyne`, so the settings is located at `~/.config/fyne/settings.json` and the preferences is located at `~/.config/fyne/${APPID}/preferences.json`. For the preferences, it's OK. But it is a question for settings, because all the `Fyne` applications use the same settings, that's the theme.

For instance, John has developed a client named `foo` with `Fyne` and David has developed a client named `bar` with `Fyne`, but they are unrelated. I have installed the two clients, `foo` and `bar`. I want that `foo` use the light theme, and that `bar` uses the dark theme. But it is impossible because they must use the same `settings.json`. I hope that `foo` and `bar` have their own `settings.json` to set the different theme. There are two solutions as follow:
1. Allow the user to modify the root configuration directory.
2. Add the APPID prefix for `settings.json`, just like `preferences.json`.

For our case, however, we need known where the root configuration directory is, and propose the first.

**FYI**

**Comment by andydotxyz** [comment]
What you described above seems more about accessing the root location property - setting it to a specific value doesn't seem like the best way to achieve that.
Is the requirement to actually change it for some reason, or is just having an API to access it all you need?

Global preferences are designed for consistency in mind. This is how most desktop systems work - if you want to override you can use the FYNE_THEME environment for specific apps.
Otherwise apps will continue to match the system default or user preference.

**Comment by xgfone** [comment]
Fine! If there is an API to access the root configuration directory, it's better.

**Comment by andydotxyz** [comment]
Great, I will change this to exporting the app root configuration directory and work on it.

**Comment by andydotxyz** [comment]
Will be released in 1.4 as `App.Storage().RootURI()`.
